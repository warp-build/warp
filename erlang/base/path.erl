-module(path).

-export([add_extension/2]).
-export([contains/2]).
-export([dirname/1]).
-export([ensure/1]).
-export([extension/1]).
-export([filename/1]).
-export([from_list/1]).
-export([is_prefix/2]).
-export([join/2]).
-export([new/1]).
-export([nth_tail/2]).
-export([relativize/1]).
-export([strip_prefix/2]).
-export([tail/1]).
-export([to_list/1]).
-export([to_string/1]).

-export_type([t/0]).

-opaque t() :: binary().

new(Str) -> binary:list_to_bin([Str]).

-spec is_prefix(Prefix :: t(), Path :: t()) -> bool().
is_prefix(Prefix, Path) -> strip_prefix(Prefix, Path) =/= Path.

-spec nth_tail(Path :: t(), N :: non_neg_integer()) -> t().
nth_tail(Path, N) when N > 0 ->
  Parts = path:to_list(Path),
  path:from_list(lists:nthtail(length(Parts) - N, Parts)).

-spec strip_prefix(Prefix :: t(), Path :: t()) -> t().
strip_prefix(Prefix, Path) ->
  Replaced = str:new(string:replace(Path, Prefix, "")),
  Final = case str:begins_with(Replaced, "/") of
            true -> str:new(string:replace(Replaced, "/", ""));
            false -> Replaced
          end,
  new([ Final ]).

join(A, B) -> from_list([A, B]).

extension(Path) -> new(filename:extension(Path)).

add_extension(Path, Ext) -> new([Path, ".", Ext]).

filename(Path) -> new(filename:basename(Path)).

dirname(Path) -> new(filename:dirname(Path)).

contains(Path, Str) -> not (string:find(Path, Str) == nomatch).

to_list(Path) -> [ new(Seg) || Seg <- filename:split(Path) ].

to_string(Path) -> binary:bin_to_list(Path).

from_list([]) -> new("");
from_list(Ls) when is_list(Ls) -> new(filename:join(Ls)).

-spec relativize(Path :: t()) -> t().
relativize(Path) ->
  case to_list(Path) of
    [<<".">> | Rest] -> from_list(Rest);
    Parts -> from_list(Parts)
  end.

-spec tail(Path :: t()) -> t().
tail(Path) -> filename:join(tl(filename:split(Path))).

-spec ensure(Path :: t()) -> ok.
ensure(Path) -> filelib:ensure_dir(Path).
