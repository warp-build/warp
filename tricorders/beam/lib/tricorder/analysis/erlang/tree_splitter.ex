defmodule Tricorder.Analysis.Erlang.TreeSplitter do
  require Logger
  alias Tricorder.Analysis.Erlang
  alias Tricorder.Analysis.TestMatcher

  def find_subtrees(file, %{include_paths: include_paths}, matcher) do
    with {:ok, ast} <- Erlang.Ast.parse(file, include_paths),
         subtrees <- subtree(ast, matcher) |> clean_subtrees do
      Logger.info("Final tree: #{inspect(subtrees, pretty: true, limit: 1_000_000)}")
      {:ok, subtrees}
    else
      diag = {:missing_dependencies, _} -> {:ok, diag}
    end
  end

  def clean_subtrees(subtrees) do
    subtrees
    |> Enum.map(fn subtree ->
      %{subtree | ast: Erlang.Ast.CleanLoc.clean(subtree.ast)}
    end)
  end

  def subtree(ast, matcher) do
    find_symbols(ast, matcher)
    |> Enum.map(fn {name, symbol} ->
      %{
        name: name,
        ast: slice_tree(ast, symbol, name, matcher)
      }
    end)
  end

  def find_symbols(ast, matcher) do
    Logger.debug("Searching for symbol: #{inspect(matcher)}")

    :erl_visitor.walk(
      ast,
      _acc = [],
      fn
        ast = {:function, _loc, name, arity, _body}, acc ->
          if TestMatcher.matches?(matcher, name) do
            Logger.debug("Found symbol: #{name}/#{arity}")
            [{{name, arity}, ast} | acc]
          else
            if {name, arity} == matcher do
              Logger.debug("Found symbol: #{name}/#{arity}")
              [{{name, arity}, ast} | acc]
            else
              Logger.debug("Skipped symbol: #{name}/#{arity}")
              acc
            end
          end

        _ast, acc ->
          acc
      end
    )
  end

  def slice_tree(ast, symbol, name, matcher) do
    Logger.debug("Slicing tree for #{inspect(name)} with symbol: #{inspect(symbol)}")
    deps = symbol_deps(ast, symbol)
    Logger.debug("Stripping tree...")
    strip_tree(ast, [name | deps], [])
  end

  def strip_tree([], _symbols, acc), do: acc |> Enum.reverse()

  def strip_tree([{:attribute, loc, :export, exports} | ast], symbols, acc) do
    Logger.debug("Found exports: #{inspect(exports, pretty: true)}")
    exports = exports |> Enum.filter(fn export -> export in symbols end)

    if Enum.empty?(exports) do
      strip_tree(ast, symbols, acc)
    else
      strip_tree(ast, symbols, [{:attribute, 0, :export, exports} | acc])
    end
  end

  def strip_tree([node = {:attribute, _, _, _} | ast], symbols, acc) do
    strip_tree(ast, symbols, [node | acc])
  end

  def strip_tree([node = {:function, _loc, name, arity, _body} | ast], symbols, acc) do
    if {name, arity} in symbols do
      # io:format("found function: ~p/~p\n", [name, arity])
      strip_tree(ast, symbols, [node | acc])
    else
      # io:format("skipped function: ~p/~p\n", [name, arity])
      strip_tree(ast, symbols, acc)
    end
  end

  def strip_tree([node | ast], symbols, acc) do
    strip_tree(ast, symbols, [node | acc])
  end

  def symbol_deps(ast, symbol) do
    deps = extract_symbol_deps(symbol)

    rec_deps =
      for dep <- deps do
        Logger.debug("Subdeps for #{inspect(dep)}")

        find_symbols(ast, dep)
        |> Enum.map(fn dep_ast -> symbol_deps(ast, dep_ast) end)
      end

    deps = List.flatten([deps | rec_deps])
    Logger.debug("Symbol dependencies: #{inspect(deps)}")
    deps
  end

  def extract_symbol_deps(ast) do
    :erl_visitor.walk(
      ast,
      _acc = [],
      fn
        _ast = {:call, _Loc, {:atom, _Loc2, name}, args}, acc ->
          [{name, length(args)} | acc]

        _ast, acc ->
          acc
      end
    )
  end
end
