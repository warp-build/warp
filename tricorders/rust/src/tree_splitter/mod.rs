use crate::ast::*;
use crate::dependency::*;
use crate::models::Symbol;

pub struct TreeSplitter {}

impl TreeSplitter {
    pub fn get_deps_all(sources: &str) -> (Vec<String>, Vec<String>) {
        let ast = syn::parse_file(sources).unwrap();
        let mut mods: Vec<String> = Vec::new();
        let mut crates: Vec<String> = Vec::new();
        let mut visitor = AllDependency {
            mods: &mut mods,
            crates: &mut crates,
        };
        syn::visit::visit_file(&mut visitor, &ast);
        (mods, crates)
    }

    pub fn tree_split(symbol: Symbol, sources: &str) -> (syn::File, String) {
        match symbol {
            Symbol::Named { name } => {
                let ast = syn::parse_file(sources).unwrap();
                let symbol_ast = Self::get_ast_named(&name, ast.clone());
                let mut deps = Self::get_interested_symbols(symbol_ast, ast.clone());
                deps.push(name);
                let stripped_ast = Self::strip_ast(ast, deps);
                let formatted_ast = prettyplease::unparse(&stripped_ast);
                (stripped_ast, formatted_ast)
            }
            _ => panic!("Expected Symbol::Named, but got something else."),
        }
    }

    pub fn strip_ast(mut file_ast: syn::File, deps: Vec<String>) -> syn::File {
        let mut ast_filter_visitor = AstFilter { deps: &deps };
        syn::visit_mut::visit_file_mut(&mut ast_filter_visitor, &mut file_ast);
        file_ast
    }

    pub fn get_interested_symbols(ast: syn::File, full_ast: syn::File) -> Vec<String> {
        let mut symbols = Vec::new();
        let mut symbols_str = Vec::new();
        let mut visitor = SymbolDependency {
            symbols: &mut symbols,
            symbols_str: &mut symbols_str,
        };
        syn::visit::visit_file(&mut visitor, &ast);

        let mut symbols_temp = symbols_str.clone();
        let mut return_symbols = symbols_str.clone();
        while !symbols_temp.is_empty() {
            let mut symbols_rec = Vec::new();
            let mut symbols_str_rec = Vec::new();
            for dep in symbols_temp.iter() {
                let symbol_ast = Self::get_ast_named(dep, full_ast.clone());
                let mut visitor_temp = SymbolDependency {
                    symbols: &mut symbols_rec,
                    symbols_str: &mut symbols_str_rec,
                };
                syn::visit::visit_file(&mut visitor_temp, &symbol_ast);
            }
            symbols_temp = symbols_str_rec.clone();
            return_symbols.append(&mut symbols_str_rec);
        }
        return_symbols
    }

    pub fn get_ast_named(symbol: &str, file: syn::File) -> syn::File {
        let subtree_option: Vec<syn::Item> = file
            .items
            .iter()
            .filter_map(|item| {
                let found_symbol = Self::get_symbol(symbol, item.clone());
                match &found_symbol {
                    Some(_) => found_symbol,
                    None => None,
                }
            })
            .collect();

        match subtree_option.first() {
            Some(x) => syn::File {
                shebang: file.shebang,
                items: vec![x.clone()],
                attrs: file.attrs,
            },
            None => syn::File {
                shebang: file.shebang,
                items: vec![],
                attrs: file.attrs,
            },
        }
    }

    pub fn get_symbol(symbol: &str, item: syn::Item) -> Option<syn::Item> {
        match item.clone() {
            syn::Item::Fn(item_fn) => {
                if item_fn.sig.ident == symbol {
                    return Some(syn::Item::Fn(item_fn));
                }
            }
            syn::Item::Const(item_const) => {
                if item_const.ident == symbol {
                    return Some(syn::Item::Const(item_const));
                }
            }
            syn::Item::Enum(item_enum) => {
                if item_enum.ident == symbol {
                    return Some(syn::Item::Enum(item_enum));
                }
            }
            syn::Item::Impl(item_impl) => {
                for impl_item in &item_impl.items {
                    match &impl_item {
                        syn::ImplItem::Const(const_impl_item) => {
                            if const_impl_item.ident == symbol {
                                return Some(item);
                            }
                        }
                        syn::ImplItem::Method(method_item) => {
                            if method_item.sig.ident == symbol {
                                return Some(item);
                            }
                        }
                        syn::ImplItem::Type(type_item) => {
                            if type_item.ident == symbol {
                                return Some(item);
                            }
                        }
                        _ => println!("Unsupported implementation item"),
                    }
                }
            }
            syn::Item::Macro2(item_macro2) => {
                if item_macro2.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Mod(item_mod) => {
                if item_mod.ident == symbol {
                    return Some(item);
                }
                for mod_item in item_mod.content.unwrap_or_default().1 {
                    let rec_symbol = Self::get_symbol(symbol, mod_item.clone());
                    if let Some(x) = rec_symbol {
                        return Some(x);
                    }
                }
            }
            syn::Item::Static(item_static) => {
                if item_static.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Struct(item_struct) => {
                if item_struct.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Trait(item_trait) => {
                if item_trait.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::TraitAlias(item_trait_alias) => {
                if item_trait_alias.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Type(item_type) => {
                if item_type.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Union(item_union) => {
                if item_union.ident == symbol {
                    return Some(item);
                }
            }
            _ => println!("Unsupported item"),
        }
        None
    }

    // Mainly used for testing purposes
    pub fn get_ast(sources: &str) -> syn::File {
        syn::parse_file(sources).unwrap()
    }
}
