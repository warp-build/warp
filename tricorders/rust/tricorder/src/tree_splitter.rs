use super::*;
pub struct TreeSplitter {}

impl TreeSplitter {
    pub fn tree_split(symbol: &str, sources: &str) -> (syn::File, String) {
        let ast = syn::parse_file(sources).unwrap();
        let symbol_ast = Self::get_ast_named(symbol, ast.clone());
        let mut deps = Self::get_interested_symbols(symbol_ast.clone(), ast.clone());
        deps.push(symbol.to_string());
        let stripped_ast = Self::strip_ast(ast, deps);
        let formatted_ast = prettyplease::unparse(&stripped_ast);
        (stripped_ast, formatted_ast)
    }

    pub fn strip_ast(ast: syn::File, deps: Vec<String>) -> syn::File {
        let mut file_ast = ast.clone();
        let mut ast_filter_visitor = AstFilter { deps: &deps };
        syn::visit_mut::visit_file_mut(&mut ast_filter_visitor, &mut file_ast);
        file_ast
    }

    pub fn get_interested_symbols(ast: syn::File, full_ast: syn::File) -> Vec<String> {
        let mut symbols = Vec::new();
        let mut symbols_str = Vec::new();
        let mut visitor = InternalDependencyAccumulator {
            symbols: &mut symbols,
            symbols_str: &mut symbols_str,
        };
        syn::visit::visit_file(&mut visitor, &ast);

        let mut symbols_temp = symbols_str.clone();
        let mut return_symbols = symbols_str.clone();
        while symbols_temp.len() != 0 {
            let mut symbols_rec = Vec::new();
            let mut symbols_str_rec = Vec::new();
            for dep in symbols_temp.iter() {
                let symbol_ast = Self::get_ast_named(dep, full_ast.clone());
                let mut visitor_temp = InternalDependencyAccumulator {
                    symbols: &mut symbols_rec,
                    symbols_str: &mut symbols_str_rec,
                };
                syn::visit::visit_file(&mut visitor_temp, &symbol_ast);
            }
            symbols_temp = symbols_str_rec.clone();
            return_symbols.append(&mut symbols_str_rec);
        }
        return_symbols
    }

    pub fn get_ast_named(symbol: &str, file: syn::File) -> syn::File {
        let subtree_option: Vec<syn::Item> = file
            .items
            .iter()
            .filter_map(|item| {
                let found_symbol = Self::get_symbol(symbol, item.clone());
                match &found_symbol {
                    Some(_) => found_symbol,
                    None => None,
                }
            })
            .collect();

        match subtree_option.first() {
            Some(x) => syn::File {
                shebang: file.shebang,
                items: vec![x.clone()],
                attrs: file.attrs,
            },
            None => syn::File {
                shebang: file.shebang,
                items: vec![],
                attrs: file.attrs,
            },
        }
    }

    pub fn get_symbol(symbol: &str, item: syn::Item) -> Option<syn::Item> {
        match item.clone() {
            syn::Item::Fn(item_fn) => {
                if item_fn.sig.ident.to_string() == symbol {
                    return Some(syn::Item::Fn(item_fn));
                }
            }
            syn::Item::Const(item_const) => {
                if item_const.ident.to_string() == symbol {
                    return Some(syn::Item::Const(item_const));
                }
            }
            syn::Item::Enum(item_enum) => {
                if item_enum.ident.to_string() == symbol {
                    return Some(syn::Item::Enum(item_enum));
                }
            }
            syn::Item::Impl(item_impl) => {
                for impl_item in &item_impl.items {
                    match &impl_item {
                        syn::ImplItem::Const(const_impl_item) => {
                            if const_impl_item.ident.to_string() == symbol {
                                return Some(item);
                            }
                        }
                        syn::ImplItem::Method(method_item) => {
                            if method_item.sig.ident.to_string() == symbol {
                                return Some(item);
                            }
                        }
                        syn::ImplItem::Type(type_item) => {
                            if type_item.ident.to_string() == symbol {
                                return Some(item);
                            }
                        }
                        _ => println!("Unsupported implementation item"),
                    }
                }
            }
            syn::Item::Macro2(item_macro2) => {
                if item_macro2.ident.to_string() == symbol {
                    return Some(item);
                }
            }
            syn::Item::Mod(item_mod) => {
                if item_mod.ident.to_string() == symbol {
                    return Some(item);
                }
                for mod_item in item_mod.content.unwrap_or_default().1 {
                    let rec_symbol = Self::get_symbol(symbol, mod_item.clone());
                    match rec_symbol {
                        Some(x) => return Some(x),
                        None => {}
                    }
                }
            }
            syn::Item::Static(item_static) => {
                if item_static.ident.to_string() == symbol {
                    return Some(item);
                }
            }
            syn::Item::Struct(item_struct) => {
                if item_struct.ident.to_string() == symbol {
                    return Some(item);
                }
            }
            syn::Item::Trait(item_trait) => {
                if item_trait.ident.to_string() == symbol {
                    return Some(item);
                }
            }
            syn::Item::TraitAlias(item_trait_alias) => {
                if item_trait_alias.ident.to_string() == symbol {
                    return Some(item);
                }
            }
            syn::Item::Type(item_type) => {
                if item_type.ident.to_string() == symbol {
                    return Some(item);
                }
            }
            syn::Item::Union(item_union) => {
                if item_union.ident.to_string() == symbol {
                    return Some(item);
                }
            }
            _ => println!("Unsupported item"),
        }
        None
    }
}
