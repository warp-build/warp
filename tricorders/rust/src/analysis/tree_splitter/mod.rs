use std::collections::HashSet;
use std::path::PathBuf;
use std::process::Command;

use crate::analysis::ast::*;
use crate::analysis::dependency::*;
pub struct TreeSplitter {}

impl TreeSplitter {
    pub fn expand_file(file_path: PathBuf) -> String {
        let cmd_out = Command::new("rustc")
            .arg("-Z")
            .arg("unpretty=expanded")
            .arg("--test")
            .arg(file_path)
            .output()
            .unwrap();

        String::from_utf8(cmd_out.stdout).unwrap()
    }

    pub fn parse_file(file_path: PathBuf) -> String {
        let cmd_out = Command::new("rustc")
            .arg("-Z")
            .arg("unpretty=normal")
            .arg("--test")
            .arg(file_path)
            .output()
            .unwrap();

        String::from_utf8(cmd_out.stdout).unwrap()
    }

    pub fn get_deps_all(sources: &str) -> (Vec<String>, Vec<String>) {
        let ast = syn::parse_file(sources).unwrap();
        let mut mods: Vec<String> = Vec::new();
        let mut crates: Vec<String> = Vec::new();
        let mut visitor = AllDependency {
            mods: &mut mods,
            crates: &mut crates,
        };
        syn::visit::visit_file(&mut visitor, &ast);
        (mods, crates)
    }

    pub fn tree_split(test_name: String, sources: String) -> (syn::File, String) {
        let ast = syn::parse_file(&sources).unwrap();
        let symbol_ast = Self::get_ast_named(&test_name, ast.clone());
        let mut deps = Self::get_interested_symbols(symbol_ast, ast.clone());
        deps.push(test_name);
        let stripped_ast = Self::strip_ast(ast, deps);
        let formatted_ast = prettyplease::unparse(&stripped_ast);
        (stripped_ast, formatted_ast)
    }

    pub fn strip_ast(mut file_ast: syn::File, deps: Vec<String>) -> syn::File {
        let mut ast_filter_visitor = AstFilter { deps: &deps };
        syn::visit_mut::visit_file_mut(&mut ast_filter_visitor, &mut file_ast);
        file_ast
    }

    pub fn has_tests(src: &str) -> bool {
        let ast = syn::parse_file(src).unwrap();
        let mut visitor = TestVisitor { has_test: false };
        syn::visit::visit_file(&mut visitor, &ast);
        visitor.has_test
    }

    pub fn find_matching_tests(test_matcher: Vec<String>, sources: &str) -> Vec<String> {
        let ast = syn::parse_file(sources).unwrap();
        let mut matching_tests: Vec<String> = Vec::new();
        if test_matcher.is_empty() {
            let mut visitor = TestFinder {
                matcher: "".to_string(),
                tests: vec![],
            };
            syn::visit::visit_file(&mut visitor, &ast);
            matching_tests.extend(visitor.tests);
        } else {
            for matcher in test_matcher.into_iter() {
                let mut visitor = TestFinder {
                    matcher,
                    tests: vec![],
                };
                syn::visit::visit_file(&mut visitor, &ast);
                matching_tests.extend(visitor.tests);
            }
        }
        matching_tests
    }

    pub fn get_interested_symbols(ast: syn::File, full_ast: syn::File) -> Vec<String> {
        let mut symbols = HashSet::new();
        let mut visitor = SymbolDependency::new(&mut symbols);
        syn::visit::visit_file(&mut visitor, &ast);

        let mut symbols_temp = symbols.clone();
        while !symbols_temp.is_empty() {
            let mut symbols_rec = HashSet::new();
            for dep in symbols_temp.iter() {
                let symbol_ast = Self::get_ast_named(dep.to_string().as_ref(), full_ast.clone());
                let mut visitor_temp = SymbolDependency::new(&mut symbols_rec);
                syn::visit::visit_file(&mut visitor_temp, &symbol_ast);
            }
            symbols_temp = symbols_rec.clone();
            symbols.extend(symbols_rec);
        }
        let mut res: Vec<String> = symbols.iter().map(|sym| sym.to_string()).collect();
        res.sort();
        res
    }

    // Returns an AST with only the needed function as the only element in the items
    pub fn get_ast_named(symbol: &str, file: syn::File) -> syn::File {
        let subtree_option: Vec<syn::Item> = file
            .items
            .iter()
            .filter_map(|item| {
                let found_symbol = Self::get_symbol(symbol, item.clone());
                match &found_symbol {
                    Some(_) => found_symbol,
                    None => None,
                }
            })
            .collect();

        match subtree_option.first() {
            Some(x) => syn::File {
                shebang: file.shebang,
                items: vec![x.clone()],
                attrs: file.attrs,
            },
            None => syn::File {
                shebang: file.shebang,
                items: vec![],
                attrs: file.attrs,
            },
        }
    }

    pub fn get_symbol(symbol: &str, item: syn::Item) -> Option<syn::Item> {
        match item.clone() {
            syn::Item::Fn(item_fn) => {
                if item_fn.sig.ident == symbol {
                    return Some(syn::Item::Fn(item_fn));
                }
            }
            syn::Item::Enum(item_enum) => {
                if item_enum.ident == symbol {
                    return Some(syn::Item::Enum(item_enum));
                }
            }
            syn::Item::Impl(item_impl) => {
                for impl_item in &item_impl.items {
                    match &impl_item {
                        syn::ImplItem::Const(const_impl_item) => {
                            if const_impl_item.ident == symbol {
                                return Some(item);
                            }
                        }
                        syn::ImplItem::Method(method_item) => {
                            if method_item.sig.ident == symbol {
                                return Some(item);
                            }
                        }
                        syn::ImplItem::Type(type_item) => {
                            if type_item.ident == symbol {
                                return Some(item);
                            }
                        }
                        _ => {}
                    }
                }
            }
            syn::Item::Macro2(item_macro2) => {
                if item_macro2.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Mod(item_mod) => {
                if item_mod.ident == symbol {
                    return Some(item);
                }
                for mod_item in item_mod.content.unwrap_or_default().1 {
                    let rec_symbol = Self::get_symbol(symbol, mod_item.clone());
                    if let Some(x) = rec_symbol {
                        return Some(x);
                    }
                }
            }
            syn::Item::Static(item_static) => {
                if item_static.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Struct(item_struct) => {
                if item_struct.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Trait(item_trait) => {
                if item_trait.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::TraitAlias(item_trait_alias) => {
                if item_trait_alias.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Type(item_type) => {
                if item_type.ident == symbol {
                    return Some(item);
                }
            }
            syn::Item::Union(item_union) => {
                if item_union.ident == symbol {
                    return Some(item);
                }
            }
            _ => {}
        }
        None
    }

    // Mainly used for testing purposes
    pub fn get_ast(sources: &str) -> syn::File {
        syn::parse_file(sources).unwrap()
    }
}
