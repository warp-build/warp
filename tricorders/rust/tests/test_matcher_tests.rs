use std::path::PathBuf;

use tricorder::analysis::tree_splitter::TreeSplitter;

#[test]
fn t001_find_matching_tests() {
    let source = TreeSplitter::expand_file(PathBuf::from("./tests/sources/t003_lib_with_tests.rs"));
    insta::assert_display_snapshot!(format!("{:#?}", &source));

    let test_name = vec!["add".to_string()];

    let matching_tests = TreeSplitter::find_matching_tests(test_name, &source);
    insta::assert_display_snapshot!(format!("{:#?}", &matching_tests));
}

#[test]
fn t002_find_all_tests() {
    let source = TreeSplitter::expand_file(PathBuf::from("./tests/sources/t003_lib_with_tests.rs"));

    let test_name = vec![];

    let matching_tests = TreeSplitter::find_matching_tests(test_name, &source);
    insta::assert_display_snapshot!(format!("{:#?}", &matching_tests));
}
