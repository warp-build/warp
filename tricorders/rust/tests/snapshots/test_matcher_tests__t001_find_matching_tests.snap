---
source: tricorders/rust/tests/test_matcher_tests.rs
expression: "format!(\"{:#?}\", & source)"
---
"#![feature(prelude_import)]\n#![no_std]\n#[prelude_import]\nuse ::std::prelude::rust_2015::*;\n#[macro_use]\nextern crate std;\npub fn add_two(a: i32) -> i32 { a + 2 }\n\n#[cfg(test)]\nmod tests {\n    use super::*;\n\n    extern crate test;\n    #[cfg(test)]\n    #[rustc_test_marker = \"tests::add_two_and_two\"]\n    pub const add_two_and_two: test::TestDescAndFn =\n        test::TestDescAndFn {\n            desc: test::TestDesc {\n                name: test::StaticTestName(\"tests::add_two_and_two\"),\n                ignore: false,\n                ignore_message: ::core::option::Option::None,\n                source_file: \"./tests/sources/t003_lib_with_tests.rs\",\n                start_line: 10usize,\n                start_col: 8usize,\n                end_line: 10usize,\n                end_col: 23usize,\n                compile_fail: false,\n                no_run: false,\n                should_panic: test::ShouldPanic::No,\n                test_type: test::TestType::Unknown,\n            },\n            testfn: test::StaticTestFn(||\n                    test::assert_test_result(add_two_and_two())),\n        };\n    fn add_two_and_two() {\n\n\n\n\n\n        match (&4, &add_two(2)) {\n            (left_val, right_val) => {\n                if !(*left_val == *right_val) {\n                        let kind = ::core::panicking::AssertKind::Eq;\n                        ::core::panicking::assert_failed(kind, &*left_val,\n                            &*right_val, ::core::option::Option::None);\n                    }\n            }\n        };\n    }\n    extern crate test;\n    #[cfg(test)]\n    #[rustc_test_marker = \"tests::add_three_and_two\"]\n    pub const add_three_and_two: test::TestDescAndFn =\n        test::TestDescAndFn {\n            desc: test::TestDesc {\n                name: test::StaticTestName(\"tests::add_three_and_two\"),\n                ignore: false,\n                ignore_message: ::core::option::Option::None,\n                source_file: \"./tests/sources/t003_lib_with_tests.rs\",\n                start_line: 15usize,\n                start_col: 8usize,\n                end_line: 15usize,\n                end_col: 25usize,\n                compile_fail: false,\n                no_run: false,\n                should_panic: test::ShouldPanic::No,\n                test_type: test::TestType::Unknown,\n            },\n            testfn: test::StaticTestFn(||\n                    test::assert_test_result(add_three_and_two())),\n        };\n    fn add_three_and_two() {\n        match (&5, &add_two(3)) {\n            (left_val, right_val) => {\n                if !(*left_val == *right_val) {\n                        let kind = ::core::panicking::AssertKind::Eq;\n                        ::core::panicking::assert_failed(kind, &*left_val,\n                            &*right_val, ::core::option::Option::None);\n                    }\n            }\n        };\n    }\n    extern crate test;\n    #[cfg(test)]\n    #[rustc_test_marker = \"tests::one_hundred\"]\n    pub const one_hundred: test::TestDescAndFn =\n        test::TestDescAndFn {\n            desc: test::TestDesc {\n                name: test::StaticTestName(\"tests::one_hundred\"),\n                ignore: false,\n                ignore_message: ::core::option::Option::None,\n                source_file: \"./tests/sources/t003_lib_with_tests.rs\",\n                start_line: 20usize,\n                start_col: 8usize,\n                end_line: 20usize,\n                end_col: 19usize,\n                compile_fail: false,\n                no_run: false,\n                should_panic: test::ShouldPanic::No,\n                test_type: test::TestType::Unknown,\n            },\n            testfn: test::StaticTestFn(||\n                    test::assert_test_result(one_hundred())),\n        };\n    fn one_hundred() {\n        let res = super::add_two(100);\n        match (&102, &res) {\n            (left_val, right_val) => {\n                if !(*left_val == *right_val) {\n                        let kind = ::core::panicking::AssertKind::Eq;\n                        ::core::panicking::assert_failed(kind, &*left_val,\n                            &*right_val, ::core::option::Option::None);\n                    }\n            }\n        };\n    }\n}\nextern crate test;\n#[cfg(test)]\n#[rustc_test_marker = \"no_mod_test_1\"]\npub const no_mod_test_1: test::TestDescAndFn =\n    test::TestDescAndFn {\n        desc: test::TestDesc {\n            name: test::StaticTestName(\"no_mod_test_1\"),\n            ignore: false,\n            ignore_message: ::core::option::Option::None,\n            source_file: \"./tests/sources/t003_lib_with_tests.rs\",\n            start_line: 27usize,\n            start_col: 4usize,\n            end_line: 27usize,\n            end_col: 17usize,\n            compile_fail: false,\n            no_run: false,\n            should_panic: test::ShouldPanic::No,\n            test_type: test::TestType::Unknown,\n        },\n        testfn: test::StaticTestFn(||\n                test::assert_test_result(no_mod_test_1())),\n    };\nfn no_mod_test_1() {\n    match (&4, &add_two(2)) {\n        (left_val, right_val) => {\n            if !(*left_val == *right_val) {\n                    let kind = ::core::panicking::AssertKind::Eq;\n                    ::core::panicking::assert_failed(kind, &*left_val,\n                        &*right_val, ::core::option::Option::None);\n                }\n        }\n    };\n}\nextern crate test;\n#[cfg(test)]\n#[rustc_test_marker = \"no_mod_test_2\"]\npub const no_mod_test_2: test::TestDescAndFn =\n    test::TestDescAndFn {\n        desc: test::TestDesc {\n            name: test::StaticTestName(\"no_mod_test_2\"),\n            ignore: false,\n            ignore_message: ::core::option::Option::None,\n            source_file: \"./tests/sources/t003_lib_with_tests.rs\",\n            start_line: 32usize,\n            start_col: 4usize,\n            end_line: 32usize,\n            end_col: 17usize,\n            compile_fail: false,\n            no_run: false,\n            should_panic: test::ShouldPanic::No,\n            test_type: test::TestType::Unknown,\n        },\n        testfn: test::StaticTestFn(||\n                test::assert_test_result(no_mod_test_2())),\n    };\nfn no_mod_test_2() {\n    match (&5, &add_two(3)) {\n        (left_val, right_val) => {\n            if !(*left_val == *right_val) {\n                    let kind = ::core::panicking::AssertKind::Eq;\n                    ::core::panicking::assert_failed(kind, &*left_val,\n                        &*right_val, ::core::option::Option::None);\n                }\n        }\n    };\n}\nextern crate test;\n#[cfg(test)]\n#[rustc_test_marker = \"add_4_and_two\"]\npub const add_4_and_two: test::TestDescAndFn =\n    test::TestDescAndFn {\n        desc: test::TestDesc {\n            name: test::StaticTestName(\"add_4_and_two\"),\n            ignore: false,\n            ignore_message: ::core::option::Option::None,\n            source_file: \"./tests/sources/t003_lib_with_tests.rs\",\n            start_line: 37usize,\n            start_col: 4usize,\n            end_line: 37usize,\n            end_col: 17usize,\n            compile_fail: false,\n            no_run: false,\n            should_panic: test::ShouldPanic::No,\n            test_type: test::TestType::Unknown,\n        },\n        testfn: test::StaticTestFn(||\n                test::assert_test_result(add_4_and_two())),\n    };\nfn add_4_and_two() {\n    match (&6, &add_two(4)) {\n        (left_val, right_val) => {\n            if !(*left_val == *right_val) {\n                    let kind = ::core::panicking::AssertKind::Eq;\n                    ::core::panicking::assert_failed(kind, &*left_val,\n                        &*right_val, ::core::option::Option::None);\n                }\n        }\n    };\n}\n#[rustc_main]\npub fn main() -> () {\n    extern crate test;\n    test::test_main_static(&[&add_4_and_two, &no_mod_test_1, &no_mod_test_2,\n                    &add_three_and_two, &add_two_and_two, &one_hundred])\n}\n"
